src/v2/BaseLoginRouter.js

```diff
@@ -129,26 +81,18 @@ export default Router.extend({
   },
 
   handleIdxResponseSuccess (idxResponse) {
-    const { interactionCode } = idxResponse;
-    if (interactionCode) {
-      const authClient = this.settings.getAuthClient();
-      const codeVerifier = this.settings.get('codeVerifier');
-      this.settings.unset('codeVerifier');
-      return authClient.token.exchangeCodeForTokens({ codeVerifier, interactionCode })
-        .then(({ tokens }) => {
-          this.settings.callGlobalSuccess(Enums.SUCCESS, { tokens });
-        })
-        .catch(err => {
-          this.settings.callGlobalError(err);
-        });
+    if (idxResponse.interactionCode) {
+      return completeLoginFlow(this.settings, idxResponse);
     }
-
+  
     // transform response
-    const ionResponse = _.compose(
-      i18nTransformer,
-      uiSchemaTransformer,
-      responseTransformer.bind({}, this.settings),
-    )(idxResponse);
+    const ionResponse = transformIdxResponse(this.settings, idxResponse);
+
+    // if this is a terminal error, clear all transaction meta
+    if (IonResponseHelper.isTerminalError(ionResponse)) {
+      clearTransactionMeta(this.settings);
```

Comment:
Can we do this in `TerminalView.js`?

Reply:
I did put it there originally, following[ the pattern here](https://github.com/okta/okta-signin-widget/pull/1616/files#diff-ac1cb237384729e4493602e41ab309d0fe0504e0d9021015f1d45e195ab71acdR91).  But I decided it would be better to intercept this case earlier so that it does not depend on any logic embedded in the UI layer.

Reply:
Also I don't want to clear in ALL terminal cases, only the cases where it is an ERROR. Some terminal cases are just telling you it is redirecting... We want to keep the saved meta through that redirect.

Reply:
> Some terminal cases are just telling you it is redirecting... We want to keep the saved meta through that redirect.

curious what are those?

- SuccessView for redirecting when success
- RedirectIDP has its own view as well iirc.

My intent was to reuse the dispatch logic instead of creating another similar one.

Also just discovery there is another terminal view that isn't extend from TerminalView. `DeviceEnrollmentTerminalView`..

Reply:
We can fine-tune the logic of `IonResponseHelper.isTerminalError` to capture the desired states. What these actual cases are is still not clear at this time, there isn't any spec around this. I think it's good to have a place to put this logic since it will probably need adjustment in the future.

This logic is distinct from the router's decision to show a particular view. The idea of this function is that it should return true if the response indicates a **permanent error** and therefore any saved transaction data should be cleared to prevent re-use (and possible infinite loop)

Reply:
The Terminal is more a concept we created for UI now and it's created at `responseTransformer` where it sort of indicates permanent error plus special case like [skipping while server at read only mode](https://github.com/okta/okta-signin-widget/pull/1529)
While we reach this success handler, we won't expect empty `remediations` otherwise bug.

If we worry about mix up those special logic with view layer, we create have a similar factory like ViewFactory to create special handler for particular view so that we'll have center place to manage (dispatching) those special case. (we can refactoring existing logic for clean up session storage as well)
If you like this idea, we can chat more about it and deal with it later and use to same pattern like session storage clean up for now (unless it could trigger `possible infinite loop` which seems like not atm)

Reply:
Also I guess we have to clean up in `DeviceEnrollmentTerminalView` as well?